### PR TITLE
feat: support high priority micro task

### LINKED ISSUE #22: promised tasks
**I'm submitting a feature request**

* **Library Version:**
1.2.1

**Please tell us about your environment:**
* **Operating System:**
Linux - KDE Neon 5.12

* **Node Version:**
9.6.1

* **NPM Version:**
5.6.0

* **Yarn Version:**
1.3.2

* **JSPM OR Webpack AND Version**
webpack 4.0.1

* **Browser:**
all
* **Language:**
all

**Current behavior:**
`TaskQueue#queueMicroTask` just takes a function/`Task` and returns `void`
```ts
queueMicroTask(task: Task | Function): void;
```
**Expected/desired behavior:**
`TaskQueue#queueMicroTask` takes a function/`Task` and returns `Promise<void>` which resolves when the task is called.
```ts
queueMicroTask(task: Task | Function): Promise<void>;
```
* **What is the motivation / use case for changing the behavior?**
there are times when I need 2 separate "tasks" to happen but I need the first task's side effects (binding, `valueChanged` functions, etc) to settle down, before registering the second one's side effects.

currently the way to achieve this is to register the task that has side effects, and from inside that task register the second task.

some pseudocode:
```ts
@autoinject
class Page {
  constructor(private taskQueue: TaskQueue){}

  someMethod(){
    // wait for side effects to settle down
    this.taskQueue.queueMicroTask(() => {
      this.someProp = 5;

      // wait again
      this.taskQueue.queueMicroTask(() => {
        this.focusNextElementInForm();
      });
    });
  }
}
```
and if we wanted to have more things happen in succession, we would quickly find ourselves in callback hell. the same process written using `async`/`await` is much more readable
```ts
@autoinject
class Page {
  constructor(private taskQueue: TaskQueue){}

  async someMethod(){
    // wait for side effects to settle down
    await this.taskQueue.queueMicroTask(() => {
      this.someProp = 5;
    });
    // wait again
    await this.taskQueue.queueMicroTask(() => {
      this.focusNextElementInForm();
    });
  }
}
```

to achieve this behavior i'm overriding the `TaskQueue#queueMicroTask` method, with something along the lines of:
```ts
const oldQueueMicro = TaskQueue.prototype.queueMicroTask;
TaskQueue.prototype.queueMicroTask = function queueMicroTask(task) {
  return new Promise((resolve, reject) => {
    try {
      oldQueueMicro.call(this, () => {
        task.call();
        resolve();
      });
    } catch (e) {
      reject(e);
    }
  });
};
```
it works, but it's not the best

### PR DESCRIPTION
**Pull request by 3cp**
feat: support high priority micro task

closes #22, help resolving aurelia-contrib/aurelia-getter-throttle#1

### CODE REVIEW

**Comment by 3cp** [comment]
@bigopon have a look. I tested this with getterThrottle, it works.

**Comment by bigopon** [comment]
Lgtm. Perf hit is minor ( local var vs property access). Its @EisenbergEffect 's call

**Comment by 3cp** [comment]
I can hardly image this new api to be broadly used. But it does enable proper rendering optimisation from aurelia-contrib/aurelia-getter-throttle. It also provides extra flexibility on code execution.

Also, we might need a better name for the new api.

**Comment by 3cp** [comment]
Changed to WIP. I got an idea to refactor it, avoiding mutating microTaskQueue (splice on big array might be costly).

**Comment by 3cp** [comment]
Refactored to use dedicated microTaskJumpQueue array, avoid inserting task to microTaskQueue.

**Comment by EisenbergEffect** [comment]
Perhaps we should call it a `priorityQueue`???

**Comment by 3cp** [comment]
Priority sounds better than jumpQueue.

@EisenbergEffect questions:

1. should we normalise the api to `queueMicroTask(task: Task | Function, highPriority?: boolean): void;`?
  * when flushing, the behaviour is what's jumpQueue doing now.
  * when not flushing, should we ignore `highPriority`, or should we honour `highPriority` ~~to use `unshift(task)` instead of `push(task)`~~?
  * ~~`unshift` has a surprising effect. When queue high priority tasks A and B, the queue will end up as `[B, A, ...]` with B, A in reversed order. Another way to get around this is to let user to put a flag on task itself (`task.highPriority = true;`) before queuing. **The flag on task seems better**, because it doesn't require visible API change, it also helps to keep the order of high priority tasks~~.

**Update: just realised if I keep using dedicated `microTaskPriorityQueue` for both flushing and not-flusing situations, I would not have any issue with the order of high priority tasks.**

2. should we add support of highPriority on macro task queue too? I cannot image the usecase for it. But this makes the API consistent.

**Comment by EisenbergEffect** [comment]
I'm thinking of an API like this:

```TypeScript
enum QueuePriority {
  normal,
  high
}

queueMicroTask(task: Task | Function, priority = QueuePriority.normal);
```

**Comment by 3cp** [comment]
@bigopon @EisenbergEffect can I ask a TypeScript question? I don't know enough about it.

The setup uses babel with syntax-flow (which doesn't support enum) to write js file, then use babel-dts-generator to write d.ts file.

For instance, in `aurelia-binding`, `bindingMode` was created as `const` in js file.
```js
export const bindingMode = {
  oneTime: 0,
  toView: 1,
  oneWay: 1,
  twoWay: 2,
  fromView: 3
};
```

Then manually declared as `enum` in manual maintained `src/aurelia-binding.d.ts` (which babel-dts-generator honoured).
```ts
export declare enum bindingMode {
  oneTime = 0,
  toView = 1,
  oneWay = 1,
  twoWay = 2,
  fromView = 3
}
```

Do I need to create a `src/aurelia-task-queue.d.ts` in order to support `enum queuePriority` properly?

**Comment by EisenbergEffect** [comment]
We could convert this library to TypeScript... @bigopon What do you think?

**Comment by 3cp** [comment]
Probably need to use separate PR for that convention. If some one got time/plan to do it in near future, I am glad to wait. Otherwise, I don't mind to try.

**Comment by EisenbergEffect** [comment]
You can update this PR to export it as a const object. Then, in a future PR we can either convert to TS or create a custom d.ts file with the enum.

**Comment by bigopon** [comment]
I think with the upcoming changes in templating, we have the chance to refactor build of templating - resources, maybe it's a good chance to start from there, as it can always be treated as independent module, then apply the build process from there to here. And finally merge with @ZHollingshead work at https://github.com/aurelia/framework/pull/850 

Or we can go all in and continue the work of @ZHollingshead at https://github.com/aurelia/framework/pull/850 and merge this in as one of the initial packages

@EisenbergEffect @huochunpeng

**Comment by EisenbergEffect** [comment]
That makes sense. We can tackle the TS move as part of the monorepo work.

**Comment by 3cp** [comment]
Please review. Test code was cleaned up for eslint warnings, added test coverage for existing code too.

**Comment by EisenbergEffect** [comment]
@huochunpeng I like the idea but I'm a little nervous since changing this will affect the entire binding system's performance. @bigopon @jdanyow What do you think?

Another idea would be to experiment with implementing the queue as a linked list where normal priority puts things at the end and high priority puts them at the beginning. On that thought, could you do the same thing with an array and loop using while/pop until there's none left, rather than a for loop?

Basically, I'd really like to see some benchmarks on this and maybe a couple of different implementations since this would be a perf critical change.

**Comment by 3cp** [comment]
Got some benchmark http://jsben.ch/rOx3i

My linked queue implementation is primitive. During flushing, the linked queue doesn't preserve the inserting order of high priority tasks, the high priority tasks will be queued in reversed order. Need few extra lines to fix the order.

1. results vary across runs (probably due to condition over Math.random). With empty priorityQueue, the performance hit is within +/- 5% comparing to the base line.
2. surprisingly, on Chrome, the extra condition check on empty priorityQueue doesn't always make it slower?!
3. something wrong with my setup, my benchmark doesn't run in FireFox/Edge :-(
4. linked queue is about 5-10% faster.

Chrome
chrome

Safari
safari

**Comment by EisenbergEffect** [comment]
Definitely interesting. I'd love ot see how it compares with FF and Edge as well, if you can get that working.

**Comment by 3cp** [comment]
I had bit play on jsben, it looks like the usage of `eval()` (jsben's way to run test) is the reason why Firefox is so slow in jsben.

Firefox has unfixed performance issue (200 times slower in loading!) on `eval()`. https://bugzilla.mozilla.org/show_bug.cgi?id=1242133

This doesn't reflect Firefox Quantum's JS performance in normal usage, which is comparable to Chrome.

To make it work in Firefox, I need to drop queue size from 5000 to 10. My home machine is 8 yrs old, you probably can pump it up bit more on your machine. http://jsben.ch/koUT1

Due to the small queue size and Math.random, the results vary quite a lot across runs. It generally gave me a feeling linked queue is slower than array on Firefox. This assertion is probably not trustable due to small queue size.

When I got more time, I will probably do a handmade benchmark page, with full task-queue source, plus some alternative implementations on flushQueue.

Firefox
screen shot 2018-06-17 at 12 21 06 am

**Comment by EisenbergEffect** [comment]
Since we've found a way to remove the task queue in vNext, I'm not sure we want to enhance it's functionality prior to that in vCurrent. Closing this PR for now.
